jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/ContentSourceRange.java

```diff
@@ -0,0 +1,140 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.io.internal;
+
+import org.eclipse.jetty.io.Content;
+import org.eclipse.jetty.util.TypeUtil;
+
+public class ContentSourceRange implements Content.Source
+{
+    private final long _offset;
+    private final long _length;
+    private final Content.Source _source;
+    long _offsetRemaining;
+    long _lengthRemaining;
+
+    public ContentSourceRange(Content.Source source, long offset, long length)
+    {
+        _source = source;
+        _offset = offset;
+        _length = TypeUtil.checkOffsetLengthSize(offset, length, source.getLength());
+        _offsetRemaining = _offset;
+        _lengthRemaining = _length;
+    }
+
+    @Override
+    public long getLength()
+    {
+        return _length;
+    }
+
+    @Override
+    public Content.Chunk read()
+    {
+        while (true)
+        {
+            Content.Chunk chunk = _source.read();
+            if (chunk == null)
+                return null;
+
+            if (_offsetRemaining > 0)
+            {
+                if (_offsetRemaining >= chunk.remaining())
+                {
+                    // We can skip this whole chunk.
+                    _offsetRemaining -= chunk.remaining();
+                    if (chunk.isLast())
+                    {
+                        chunk.clear();
+                        return chunk;
+                    }
+
+                    chunk.release();
+                    continue;
+                }
+                else
+                {
+                    // Advance position to the correct offset.
+                    chunk.skip(_offsetRemaining);
+                    _offsetRemaining = 0;
+                }
+            }
+
+            // We can start processing the limited length if we have reached the starting offset.
+            if (_offsetRemaining == 0 && _lengthRemaining >= 0)
+            {
+                if (_lengthRemaining == 0)
+                {
+                    // Release the chunk and continue until we find a last chunk.
```

Comment:
@lachlan-roberts I wonder if we should just fail the source and return EOF rather than reading until the end? Maybe an option to do that?

Reply:
Yeah good point, I added this as an option (which is now the default).